modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +331,230 @@ void run(ClusterState state) {
+            assert backingIndex != null : "the data stream backing indices must exist";
+            List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                index,
+                metadata::index,
+                nowSupplier
+            );
+            if (downsamplingRounds.isEmpty()) {
+                continue;
+            }
+
+            String indexName = index.getName();
+            IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+            String backingIndexDownsamplingSource = IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME.get(backingIndex.getSettings());
+
+            // if the current index is not a downsample we want to mark the index as read-only before proceeding with downsampling
+            if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource) == false
+                && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
+                affectedIndices.add(index);
+                addIndexBlockOnce(indexName);
+                continue;
+            } else if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource)
+                && backingIndexDownsamplingStatus.equals(SUCCESS)) {
+                    // if the backing index is a downsample index itself, let's check if its source index still exists as we must delete it
+                    Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
+                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
+                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
+                        // TODO document that we don't handle downsample indices that were added to the data stream manually (because we
+                        // TODO currently can't reliably identify the source index to delete when multiple rounds of donwsampling are
+                        // TODO involved unless DSL stores the needed metadata in the index metadata)
+                        continue;
+                    }
+                    String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
+                    IndexMetadata downsampleSourceIndex = metadata.index(actualDownsamplingSource);
+                    if (downsampleSourceIndex != null) {
+                        affectedIndices.add(downsampleSourceIndex.getIndex());
+                        // delete downsampling source index (that's not part of the data stream anymore) before doing any more
+                        // downsampling
+                        deleteIndexOnce(backingIndexDownsamplingSource, "replacement with its downsampled index in the data stream");
+                        continue;
+                    }
```

Comment:
This might be a better view of the suggestion:

```diff
diff --git a/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java b/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java
index 8a569b7c0de..2648eb97525 100644
--- a/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java
+++ b/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java
@@ -390,34 +390,30 @@ public class DataStreamLifecycleService implements ClusterStateListener, Closeab
                 && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
                 affectedIndices.add(index);
                 addIndexBlockOnce(indexName);
-                continue;
             } else if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource)
                 && backingIndexDownsamplingStatus.equals(SUCCESS)) {
                     // if the backing index is a downsample index itself, let's check if its source index still exists as we must delete it
                     Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
-                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
-                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
-                        // TODO document that we don't handle downsample indices that were added to the data stream manually (because we
-                        // TODO currently can't reliably identify the source index to delete when multiple rounds of donwsampling are
-                        // TODO involved unless DSL stores the needed metadata in the index metadata)
-                        continue;
-                    }
-                    String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
-                    IndexMetadata downsampleSourceIndex = metadata.index(actualDownsamplingSource);
-                    if (downsampleSourceIndex != null) {
-                        affectedIndices.add(downsampleSourceIndex.getIndex());
-                        // delete downsampling source index (that's not part of the data stream anymore) before doing any more
-                        // downsampling
-                        deleteIndexOnce(backingIndexDownsamplingSource, "replacement with its downsampled index in the data stream");
-                        continue;
+                    // TODO document that we don't handle downsample indices that were added to the data stream manually (because we
+                    // TODO currently can't reliably identify the source index to delete when multiple rounds of donwsampling are                                                                            
+                    // TODO involved unless DSL stores the needed metadata in the index metadata)
+                    if (lifecycleMetadata != null && lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX)) {
+                        String actualDownsamplingSource = lifecycleMetadata.get(REPLACEMENT_SOURCE_INDEX);
+                        IndexMetadata downsampleSourceIndex = metadata.index(actualDownsamplingSource);
+                        if (downsampleSourceIndex != null) {
+                            affectedIndices.add(downsampleSourceIndex.getIndex());
+                            // delete downsampling source index (that's not part of the data stream anymore) before doing any more
+                            // downsampling
+                            deleteIndexOnce(backingIndexDownsamplingSource, "replacement with its downsampled index in the data stream");                                                            
+                        }
                     }
+                } else {
+                    // this index has matching downsample rounds and is read-only, so let's wait for an in-progress downsampling operation
+                    // to succeed or trigger the last matching round
+                    affectedIndices.addAll(
+                        checkEarlyRoundsForMaintenanceOrDownsampleLastRound(dataStream, backingIndex, downsamplingRounds, metadata)
+                    );
                 }
-
-            // this index has matching downsample rounds and is read-only, so let's wait for an in-progress downsampling operation
-            // to succeed or trigger the last matching round
-            affectedIndices.addAll(
-                checkEarlyRoundsForMaintenanceOrDownsampleLastRound(dataStream, backingIndex, downsamplingRounds, metadata)                                               
-            );
         }
 
         return affectedIndices;
```

Reply:
Flipped the metadata check as suggested.